src/core/Hydrogen.cpp

```diff
@@ -1453,7 +1429,7 @@ int audioEngine_process( uint32_t nframes, void* /*arg*/ )
 			+ ( finishTimeval.tv_usec - startTimeval.tv_usec ) / 1000.0;
 
 	if ( m_audioEngineState == STATE_PLAYING ) {
-		AudioEngine::get_instance()->updateElapsedTime( m_pAudioDriver->getBufferSize(),
+		AudioEngine::get_instance()->updateElapsedTime( nframes,
```

Comment:
There is https://github.com/hydrogen-music/hydrogen/blob/master/src/core/IO/PulseAudioDriver.cpp#L280 which might be the cause of #884. Haven't looked into PulseAudio yet and don't know if the change is something bad or rather a cure.

Reply:
Oh boy, that's a weird one.

So, that `std::min` call looks as if what it's *trying* to do is account for the possibility of PulseAudio asking for more bytes than a single buffer will hold. It's not actually doing that though because

 1. I don't think PulseAudio would actually do that, because it's been told that's the *max* buffer size
 2. It doesn't actually push the data out to PulseAudio for each call to `audioEngine_process` -- it just chunks up generation of the data, and 
 3. it takes the minimum of the number of *frames* requested, and the max number of *bytes* of the buffer. Since PulseAudio should already be ensuring the number of frames will fit in the buffer, and the number of bytes is 4 times the number of frames, the min should always just be `num_samples`.

I don't think it's the actual *cause* of either of the issues in #884 though, since I don't think it ever does anything at all. For part 2, I think I will remove that informational message as part of this change -- because it wall happen *all* the time under CoreAudio with mismatched device/AudioUnit sample rates, and it looks like PulseAudio is entirely intentionally passing variable numbers of samples to the write callback (this makes perfect sense for an audio system that may depend on network for data). In general we just shouldn't expect a fixed size buffer. The only trouble remains the use of buffer size as a proxy for latency. Hmm.

Reply:
> The only trouble remains the use of buffer size as a proxy for latency

The `AudioEngine::updateElapsedTime()` method is only used to display the correct time that passed since the start of the playback in the Player Control. I introduced this because the former version based on the tick size showed nonsense when either a tempo marker was passed or the tempo/position is being controlled by the JACK server. It should have no problems with dynamic sizes at all. It only cares about counting the frames passed since playback started.

Reply:
Yes, that use is fine -- changing it to nFrames there is, I think, exactly the right thing to do; there's another use somewhere else I meant. Hold on... [this one](https://github.com/hydrogen-music/hydrogen/blob/11b32c45f8fc88fee2423286b204c390d764462c/src/core/Hydrogen.cpp#L2775).

I think the right thing to do with this is to remove `getBufferSize` from the driver interface and add a method that returns an approximate latency. PulseAudio has an interface to get this, and drivers which use a constant buffer size can just return that by default.